local ExtraTypes = {}

export type callbackZone = (Player: Player, Zone: PlayerZoneConfig) -> ()

export type PromiseLike = {
	andThen: (self: PromiseLike, onFulfilled: Opt<AnyFn>, onRejected: Opt<AnyFn>) -> PromiseLike,
	catch: (self: PromiseLike, onRejected: AnyFn) -> PromiseLike,
	finally: (self: PromiseLike, onFinally: Fn<>) -> PromiseLike,
	tap: (self: PromiseLike, onTap: (any) -> ()) -> PromiseLike,
	andThenCall: (self: PromiseLike, callback: AnyFn, ...any) -> PromiseLike,
	andThenReturn: (self: PromiseLike, ...any) -> PromiseLike,
	finallyCall: (self: PromiseLike, callback: AnyFn, ...any) -> PromiseLike,
	finallyReturn: (self: PromiseLike, ...any) -> PromiseLike,
	now: (self: PromiseLike, rejectionValue: any?) -> PromiseLike,
	timeout: (self: PromiseLike, seconds: number, rejectionValue: any?) -> PromiseLike,

	cancel: (self: PromiseLike) -> (),
	getStatus: (self: PromiseLike) -> Status,
	awaitStatus: (self: PromiseLike) -> (Status, any?),
	expect: (self: PromiseLike) -> any,
	await: (self: PromiseLike) -> (boolean, any),
}

export type R6 = {
	Head: {
		Mesh: SpecialMesh,
		face: Decal,
		HairAttachment: Attachment,
		HatAttachment: Attachment,
		FaceFrontAttachment: Attachment,
		FaceCenterAttachment: Attachment,
	} & Part,
	Torso: {
		roblox: Decal,
		NeckAttachment: Attachment,
		BodyFrontAttachment: Attachment,
		BodyBackAttachment: Attachment,
		LeftCollarAttachment: Attachment,
		RightCollarAttachment: Attachment,
		WaistFrontAttachment: Attachment,
		WaistCenterAttachment: Attachment,
		WaistBackAttachment: Attachment,
		Right_Shoulder: Motor6D,
		Left_Shoulder: Motor6D,
		Right_Hip: Motor6D,
		Left_Hip: Motor6D,
		Neck: Motor6D,
	} & Part,
	Left_Arm: {
		LeftShoulderAttachment: Attachment,
		LeftGripAttachment: Attachment,
	} & Part,
	Right_Arm: {
		RightShoulderAttachment: Attachment,
		RightGripAttachment: Attachment,
	} & Part,
	Left_Leg: {
		LeftFootAttachment: Attachment,
	} & Part,
	Right_Leg: {
		RightFootAttachment: Attachment,
	} & Part,
	Humanoid: {
		Status: Status,
		Animator: Animator,
		HumanoidDescription: {
			BodyPartDescription_1: BodyPartDescription,
			BodyPartDescription_2: BodyPartDescription,
			BodyPartDescription_3: BodyPartDescription,
			BodyPartDescription_4: BodyPartDescription,
			BodyPartDescription_5: BodyPartDescription,
			BodyPartDescription_6: BodyPartDescription,
		} & HumanoidDescription,
	} & Humanoid,
	HumanoidRootPart: {
		RootAttachment: Attachment,
		RootJoint: Motor6D,
	} & Part,
	Animate: {
		idle: {
			Animation1: {
				Weight: NumberValue,
			} & Animation,
			Animation2: {
				Weight: NumberValue,
			} & Animation,
		} & StringValue,
		walk: {
			WalkAnim: Animation,
		} & StringValue,
		run: {
			RunAnim: Animation,
		} & StringValue,
		jump: {
			JumpAnim: Animation,
		} & StringValue,
		climb: {
			ClimbAnim: Animation,
		} & StringValue,
		toolnone: {
			ToolNoneAnim: Animation,
		} & StringValue,
		fall: {
			FallAnim: Animation,
		} & StringValue,
		sit: {
			SitAnim: Animation,
		} & StringValue,
		PlayEmote: BindableFunction,
		ScaleDampeningPercent: NumberValue,
	} & LocalScript,
	Body_Colors: BodyColors,
} & Model

-- Generated by AutoType plugin
export type R15 = {
	Head: {
		Mesh: SpecialMesh,
		face: Decal,
		HairAttachment: Attachment,
		HatAttachment: Attachment,
		FaceFrontAttachment: Attachment,
		FaceCenterAttachment: Attachment,
	} & Part,
	Torso: {
		roblox: Decal,
		NeckAttachment: Attachment,
		BodyFrontAttachment: Attachment,
		BodyBackAttachment: Attachment,
		LeftCollarAttachment: Attachment,
		RightCollarAttachment: Attachment,
		WaistFrontAttachment: Attachment,
		WaistCenterAttachment: Attachment,
		WaistBackAttachment: Attachment,
		Right_Shoulder: Motor6D,
		Left_Shoulder: Motor6D,
		Right_Hip: Motor6D,
		Left_Hip: Motor6D,
		Neck: Motor6D,
	} & Part,
	Left_Arm: {
		LeftShoulderAttachment: Attachment,
		LeftGripAttachment: Attachment,
	} & Part,
	Right_Arm: {
		RightShoulderAttachment: Attachment,
		RightGripAttachment: Attachment,
	} & Part,
	Left_Leg: {
		LeftFootAttachment: Attachment,
	} & Part,
	Right_Leg: {
		RightFootAttachment: Attachment,
	} & Part,
	Humanoid: {
		Status: Status,
		Animator: Animator,
		HumanoidDescription: {
			BodyPartDescription_1: BodyPartDescription,
			BodyPartDescription_2: BodyPartDescription,
			BodyPartDescription_3: BodyPartDescription,
			BodyPartDescription_4: BodyPartDescription,
			BodyPartDescription_5: BodyPartDescription,
			BodyPartDescription_6: BodyPartDescription,
		} & HumanoidDescription,
	} & Humanoid,
	HumanoidRootPart: {
		RootAttachment: Attachment,
		RootJoint: Motor6D,
	} & Part,
	Animate: {
		idle: {
			Animation1: {
				Weight: NumberValue,
			} & Animation,
			Animation2: {
				Weight: NumberValue,
			} & Animation,
		} & StringValue,
		walk: {
			WalkAnim: Animation,
		} & StringValue,
		run: {
			RunAnim: Animation,
		} & StringValue,
		jump: {
			JumpAnim: Animation,
		} & StringValue,
		climb: {
			ClimbAnim: Animation,
		} & StringValue,
		toolnone: {
			ToolNoneAnim: Animation,
		} & StringValue,
		fall: {
			FallAnim: Animation,
		} & StringValue,
		sit: {
			SitAnim: Animation,
		} & StringValue,
		PlayEmote: BindableFunction,
		ScaleDampeningPercent: NumberValue,
	} & LocalScript,
	Body_Colors: BodyColors,
} & Model

export type Inventory = { [string]: number }

export type DataTemplate = {
	Cash: number,
	Lucky: number,

	Vip: boolean,
	DoubleCash: boolean,
	Gamepasses: { [string]: boolean },

	Item: Inventory,
	UnlockedItems: { any },
	UnlockedSkins: { any },

	FirstLogin: boolean,
	HasIncense: boolean,

	Wins: number,
	Losses: number,
	CurrentZone: string,
	LastSafeZone: string,

	TotalFights: number,
	TimeSpent: number,

	RobuxSpent: number,
	CashSpent: number,
	ChestOpen: number,
	SpinWheel: number,

	MusicVolume: number,
	SoundVolume: number,
	MusicEnabled: boolean,
	SoundEnabled: boolean,
}

export type PlayerZoneConfig = {
	MinHeight: number,
	Padding: Vector3?,
	Shape: "Block" | "Ball" | "Cylinder" | "Wedge" | "CornerWedge",
}

export type DataTemplateKey =
	"Cash"
	| "Lucky"
	| "Vip"
	| "DoubleCash"
	| "Gamepasses"
	| "Item"
	| "UnlockedItems"
	| "UnlockedSkins"
	| "FirstLogin"
	| "HasIncense"
	| "Wins"
	| "Losses"
	| "CurrentZone"
	| "LastSafeZone"
	| "TotalFights"
	| "TimeSpent"
	| "RobuxSpent"
	| "CashSpent"
	| "ChestOpen"
	| "SpinWheel"
	| "MusicVolume"
	| "SoundVolume"
	| "MusicEnabled"
	| "SoundEnabled"

export type DataKeyPayload = DataTemplate & { key: DataTemplateKey }

return ExtraTypes
